x/storage/keeper/msg_server_upgrade_storage.go

```diff
@@ -0,0 +1,111 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerr "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/jackalLabs/canine-chain/x/storage/types"
+)
+
+const (
+	// replace costPerMonth when oracle/price is ready
+	// cost below is for a gb
+	costPerMonth = 4000
+	costPerDay   = costPerMonth / 30
+
+	gb           int64 = 1_000_000_000
+	hoursInMonth       = time.Hour * 720
+)
+
+func (k Keeper) UpgradeStorage(goCtx context.Context, msg *types.MsgUpgradeStorage) (*types.MsgUpgradeStorageResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	payInfo, found := k.GetStoragePaymentInfo(ctx, msg.ForAddress)
+	if !found {
+		return nil, fmt.Errorf("can't upgrade non-existing storage, please use MsgBuyStorage")
+	}
+
+	duration, err := time.ParseDuration(msg.Duration)
+	if err != nil {
+		return nil, fmt.Errorf("duration can't be parsed: %s", err.Error())
+	}
+	duration = duration.Truncate(time.Hour)
+	if duration < hoursInMonth {
+		return nil, fmt.Errorf("cannot buy less than a month (720h)")
+	}
+	dm := duration.Truncate(hoursInMonth)
+	durationInMonth := int64(dm / hoursInMonth)
```

Comment:
Any time you're doing division like this you need to use sdk.Dec's Quo function to avoid floating point division or it could cause an app hash.